{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "96a188bc-75cc-4bfe-9dd4-c34f646d74f9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# load packages and set paths\n",
    "\n",
    "import dill\n",
    "import scanpy as sc\n",
    "import os\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import pandas\n",
    "import pyranges\n",
    "# Set stderr to null to avoid strange messages from ray\n",
    "import sys\n",
    "_stderr = sys.stderr\n",
    "null = open(os.devnull,'wb')\n",
    "\n",
    "# set paths\n",
    "work_dir = '/Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART'\n",
    "tmp_dir = '/scratch2/devoes/tmp'\n",
    "\n",
    "# set market\n",
    "biomart_host = \"http://nov2020.archive.ensembl.org/\" # be sure to check for compatible capitalization nomenclature\n",
    "\n",
    "# list of known TF\n",
    "tf_file = \"/Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART/cistarget_database/HOCOMOCOv11_geneNames.txt\" \n",
    "\n",
    "# load RNA AnnData, ATAC cistopic object, motif enrichment results\n",
    "adata = sc.read_h5ad(os.path.join(work_dir, 'RNA/adata_SCA.h5ad'))\n",
    "cistopic_obj = dill.load(open(os.path.join(work_dir, 'ATAC/cistopic_obj_48Topics.pkl'), 'rb'))\n",
    "menr = dill.load(open(os.path.join(work_dir, 'motifs_no-promoter/menr.pkl'), 'rb'))"
   ]
  },
  {
   "cell_type": "raw",
   "id": "314c2132-4124-4f0f-a9bc-ad21e94aca4b",
   "metadata": {},
   "source": [
    "Start by creating a SCENIC+ object\n",
    "    holds both RNA, ATAC, and motif data\n",
    "    meta data from RNA has prefix GEX_\n",
    "    meta data from ATAC has prefix ACC_\n",
    "    \n",
    "follow step-by-step instead of wrapper https://scenicplus.readthedocs.io/en/latest/Scenicplus_step_by_step-RTD.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "61524e64-991a-475f-bb44-b5b1ef57c1c6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-06-04 19:29:43,492 cisTopic     INFO     Imputing region accessibility\n",
      "2024-06-04 19:29:43,494 cisTopic     INFO     Impute region accessibility for regions 0-20000\n",
      "2024-06-04 19:29:44,813 cisTopic     INFO     Impute region accessibility for regions 20000-40000\n",
      "2024-06-04 19:29:46,139 cisTopic     INFO     Impute region accessibility for regions 40000-60000\n",
      "2024-06-04 19:29:47,458 cisTopic     INFO     Impute region accessibility for regions 60000-80000\n",
      "2024-06-04 19:29:48,786 cisTopic     INFO     Impute region accessibility for regions 80000-100000\n",
      "2024-06-04 19:29:50,125 cisTopic     INFO     Impute region accessibility for regions 100000-120000\n",
      "2024-06-04 19:29:51,446 cisTopic     INFO     Impute region accessibility for regions 120000-140000\n",
      "2024-06-04 19:29:52,784 cisTopic     INFO     Impute region accessibility for regions 140000-160000\n",
      "2024-06-04 19:29:54,115 cisTopic     INFO     Impute region accessibility for regions 160000-180000\n",
      "2024-06-04 19:29:55,443 cisTopic     INFO     Impute region accessibility for regions 180000-200000\n",
      "2024-06-04 19:29:56,789 cisTopic     INFO     Impute region accessibility for regions 200000-220000\n",
      "2024-06-04 19:29:58,133 cisTopic     INFO     Impute region accessibility for regions 220000-240000\n",
      "2024-06-04 19:29:59,468 cisTopic     INFO     Impute region accessibility for regions 240000-260000\n",
      "2024-06-04 19:30:00,802 cisTopic     INFO     Impute region accessibility for regions 260000-280000\n",
      "2024-06-04 19:30:02,140 cisTopic     INFO     Impute region accessibility for regions 280000-300000\n",
      "2024-06-04 19:30:03,471 cisTopic     INFO     Impute region accessibility for regions 300000-320000\n",
      "2024-06-04 19:30:04,797 cisTopic     INFO     Done!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SCENIC+ object with n_cells x n_genes = 12295 x 17669 and n_cells x n_regions = 12295 x 312253\n",
       "\tmetadata_regions:'Chromosome', 'Start', 'End', 'Width', 'cisTopic_nr_frag', 'cisTopic_log_nr_frag', 'cisTopic_nr_acc', 'cisTopic_log_nr_acc'\n",
       "\tmetadata_genes:'gene_ids', 'feature_types', 'genome', 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells'\n",
       "\tmetadata_cell:'GEX_n_genes_by_counts', 'GEX_total_counts', 'GEX_total_counts_mt', 'GEX_pct_counts_mt', 'GEX_n_genes', 'GEX_doublet_score', 'GEX_predicted_doublet', 'GEX_leiden', 'ACC_Dupl_rate', 'ACC_Unique_nr_frag_in_regions', 'ACC_Unique_nr_frag', 'ACC_FRIP', 'ACC_TSS_enrichment', 'ACC_Total_nr_frag', 'ACC_Log_total_nr_frag', 'ACC_cisTopic_nr_frag', 'ACC_cisTopic_log_nr_frag', 'ACC_cisTopic_nr_acc', 'ACC_cisTopic_log_nr_acc', 'ACC_Log_unique_nr_frag', 'ACC_Total_nr_frag_in_regions', 'ACC_barcode', 'ACC_Dupl_nr_frag', 'ACC_n_genes_by_counts', 'ACC_total_counts', 'ACC_total_counts_mt', 'ACC_pct_counts_mt', 'ACC_n_genes', 'ACC_doublet_score', 'ACC_predicted_doublet', 'ACC_leiden', 'ACC_sample_id'\n",
       "\tmenr:'CTX_topics_otsu_All', 'CTX_topics_otsu_No_promoters', 'DEM_topics_otsu_All', 'DEM_topics_otsu_No_promoters', 'CTX_topics_top_3_All', 'CTX_topics_top_3_No_promoters', 'DEM_topics_top_3_All', 'DEM_topics_top_3_No_promoters', 'CTX_DARs_All', 'CTX_DARs_No_promoters', 'DEM_DARs_All', 'DEM_DARs_No_promoters'\n",
       "\tdr_cell:'GEX_X_sca', 'GEX_X_umap'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from scenicplus.scenicplus_class import create_SCENICPLUS_object\n",
    "import numpy as np\n",
    "scplus_obj = create_SCENICPLUS_object(\n",
    "    GEX_anndata = adata.raw.to_adata(),\n",
    "    cisTopic_obj = cistopic_obj,\n",
    "    menr = menr,\n",
    "    bc_transform_func = lambda x: f'{x}-GSE202543_B16_CART_tumor' #function to convert scATAC-seq barcodes to scRNA-seq ones\n",
    ")\n",
    "scplus_obj.X_EXP = np.array(scplus_obj.X_EXP.todense())\n",
    "scplus_obj"
   ]
  },
  {
   "cell_type": "raw",
   "id": "1a964b66-97ff-408e-9e43-1f1c5fdcd1bf",
   "metadata": {},
   "source": [
    "Authors recommend filtering out lowly expressed genes and accessible regions\n",
    "    Is this proper for our data?\n",
    "    This is skipped in the general tutorial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "313af75d-6a34-4085-95ac-2002498c108f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-06-04 19:30:10,102 Preprocessing INFO     Going from 17669 genes to 12080 genes.\n",
      "2024-06-04 19:32:20,679 Preprocessing INFO     Going from 312253 regions to 242124 regions.\n"
     ]
    }
   ],
   "source": [
    "from scenicplus.preprocessing.filtering import filter_genes,filter_regions\n",
    "filter_genes(scplus_obj, min_pct = 0.5) # pct [0,100] 0.5% is must be in approx 65 cells\n",
    "filter_regions(scplus_obj, min_pct = 0.5)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "859ac924-5111-443a-8697-5bd1923c42c7",
   "metadata": {},
   "source": [
    "Create cistromes\n",
    "    cistrome - group of regulatory elements for a set of genes\n",
    "                for scenic+ this appears to be used as a term to describe the colleciton of regions within the accessibility assay that contain a binding site or motif for the TF\n",
    "    \n",
    "an opt to subset regions for cell type specific cistromes (not done here)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c991d2ed-58a3-4266-a9b4-50ba91f95fc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scenicplus.cistromes import *\n",
    "merge_cistromes(scplus_obj)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "6fff333e-298d-4e14-a124-b78a76420148",
   "metadata": {},
   "source": [
    "Infer Enhancer to Gene relationships\n",
    "    These steps use the associations of gene expression and region accessibility in linear and non-linear relationships\n",
    "    \n",
    "    1. Define search space\n",
    "        default is 150kb\n",
    "        Topologically Associated Domains (TAD) Boundaries may also be used\n",
    "            TADs are units of 3D nuclear organization. \n",
    "            The boundaries are insulatory zones that restrict interaction of cis-regulatory elements to their genes\n",
    "    2. Create Enhancer-to-Gene models\n",
    "        via correlation, random forest, gradient boosting machine\n",
    "        \n",
    "** check biomart host matches the assembly you used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a47556fb-e60c-438a-8308-f06c391494ad",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-06-05 06:41:55,131 R2G          INFO     Downloading gene annotation from biomart dataset: mmusculus_gene_ensembl\n",
      "2024-06-05 06:42:03,661 R2G          INFO     Downloading chromosome sizes from: http://hgdownload.cse.ucsc.edu/goldenPath/mm10/bigZips/mm10.chrom.sizes\n",
      "2024-06-05 06:42:04,333 R2G          INFO     Extending promoter annotation to 500 bp upstream and 500 downstream\n",
      "Warning! Start and End columns now have different dtypes: int32 and int64\n",
      "Warning! Start and End columns now have different dtypes: int32 and int64\n",
      "2024-06-05 06:42:05,320 R2G          INFO     Extending search space to:\n",
      "            \t\t\t\t\t\t250000 bp downstream of the end of the gene.\n",
      "            \t\t\t\t\t\t250000 bp upstream of the start of the gene.\n",
      "Warning! Start and End columns now have different dtypes: int32 and int64\n",
      "Warning! Start and End columns now have different dtypes: int32 and int64\n",
      "2024-06-05 06:42:09,885 R2G          INFO     Intersecting with regions.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "join: Strand data from other will be added as strand data to self.\n",
      "If this is undesired use the flag apply_strand_suffix=False.\n",
      "To turn off the warning set apply_strand_suffix to True or False.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning! Start and End columns now have different dtypes: int32 and int64\n",
      "2024-06-05 06:42:11,311 R2G          INFO     Calculating distances from region to gene\n",
      "2024-06-05 06:47:09,764 R2G          INFO     Removing DISTAL regions overlapping promoters\n",
      "2024-06-05 06:47:20,735 R2G          INFO     Imploding multiple entries per region and gene\n",
      "2024-06-05 06:57:08,364 R2G          INFO     Done!\n"
     ]
    }
   ],
   "source": [
    "# define search space\n",
    "from scenicplus.enhancer_to_gene import get_search_space, calculate_regions_to_genes_relationships, GBM_KWARGS\n",
    "get_search_space(scplus_obj,\n",
    "                 biomart_host = biomart_host,\n",
    "                 species = 'mmusculus',\n",
    "                 assembly = 'mm10',\n",
    "                 extend_tss = [500,500],\n",
    "                 remove_promoters = True,\n",
    "                 upstream = [1000, 250000], # default is 150kb. I want to extend to 250kb\n",
    "                 downstream = [1000, 250000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2ced0efa-3925-4d2f-a2b1-35952b0d8190",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-06-05 07:04:16,334 R2G          INFO     Calculating region to gene importances, using GBM method\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-06-05 07:04:21,731\tINFO worker.py:1612 -- Started a local Ray instance. View the dashboard at \u001b[1m\u001b[32m127.0.0.1:8265 \u001b[39m\u001b[22m\n",
      "initializing: 100%|█████████████████████████████████████████████████████████████| 10724/10724 [1:24:35<00:00,  2.11it/s]\n",
      "Running using 12 cores: 100%|████████████████████████████████████████████████████| 10724/10724 [00:57<00:00, 187.39it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-06-05 08:29:57,827 R2G          INFO     Took 5141.491611242294 seconds\n",
      "2024-06-05 08:29:57,829 R2G          INFO     Calculating region to gene correlation, using SR method\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-06-05 08:30:02,404\tINFO worker.py:1612 -- Started a local Ray instance. View the dashboard at \u001b[1m\u001b[32m127.0.0.1:8265 \u001b[39m\u001b[22m\n",
      "initializing: 100%|█████████████████████████████████████████████████████████████| 10724/10724 [1:03:11<00:00,  2.83it/s]\n",
      "Running using 12 cores: 100%|████████████████████████████████████████████████████| 10724/10724 [00:54<00:00, 196.33it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-06-05 09:34:10,918 R2G          INFO     Took 3853.0881016254425 seconds\n",
      "2024-06-05 09:34:24,462 R2G          INFO     Done!\n"
     ]
    }
   ],
   "source": [
    "# model\n",
    "calculate_regions_to_genes_relationships(scplus_obj,\n",
    "                    ray_n_cpu = 12,\n",
    "                    _temp_dir = tmp_dir,\n",
    "                    importance_scoring_method = 'GBM',\n",
    "                    importance_scoring_kwargs = GBM_KWARGS) # imported from package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cc7e224b-70fc-4d8e-aaa2-f8cd2f7613b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save\n",
    "import pickle\n",
    "\n",
    "with open(os.path.join(work_dir, 'scenicplus/scplus_obj_hocomocov11.pkl'), 'wb') as f:\n",
    "  pickle.dump(scplus_obj, f)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "7eb680eb-9fab-4615-937b-e70448855ee3",
   "metadata": {},
   "source": [
    "Infer Transcription Factor to Gene relationships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9403515d-df92-48c2-9be8-102d66a17a03",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-06-05 10:24:55,139\tINFO worker.py:1612 -- Started a local Ray instance. View the dashboard at \u001b[1m\u001b[32m127.0.0.1:8265 \u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-06-05 10:24:56,135 TF2G         INFO     Calculating TF to gene correlation, using GBM method\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "initializing:   0%|▏                                                                 | 31/12080 [00:07<39:49,  5.04it/s]\u001b[2m\u001b[36m(raylet)\u001b[0m Spilled 2885 MiB, 14 objects, write throughput 1000 MiB/s. Set RAY_verbose_spill_logs=0 to disable this message.\n",
      "initializing:   0%|▏                                                                 | 40/12080 [00:08<28:39,  7.00it/s]\u001b[2m\u001b[36m(raylet)\u001b[0m Spilled 5184 MiB, 25 objects, write throughput 1326 MiB/s.\n",
      "initializing:   0%|▎                                                                 | 54/12080 [00:09<13:10, 15.21it/s]\u001b[2m\u001b[36m(raylet)\u001b[0m Spilled 9812 MiB, 51 objects, write throughput 1826 MiB/s.\n",
      "initializing:   1%|▍                                                                 | 76/12080 [00:11<15:38, 12.79it/s]\u001b[2m\u001b[36m(raylet)\u001b[0m Spilled 18385 MiB, 88 objects, write throughput 2124 MiB/s.\n",
      "initializing:   1%|▊                                                                | 149/12080 [00:20<22:27,  8.86it/s]\u001b[2m\u001b[36m(raylet)\u001b[0m Spilled 33297 MiB, 157 objects, write throughput 2094 MiB/s.\n",
      "initializing:   2%|█▍                                                               | 266/12080 [00:36<16:42, 11.78it/s]\u001b[2m\u001b[36m(raylet)\u001b[0m Spilled 67215 MiB, 322 objects, write throughput 2179 MiB/s.\n",
      "initializing:   4%|██▋                                                              | 503/12080 [01:06<15:44, 12.26it/s]\u001b[2m\u001b[36m(raylet)\u001b[0m Spilled 131995 MiB, 620 objects, write throughput 2144 MiB/s.\n",
      "initializing:   8%|█████▎                                                          | 1014/12080 [02:11<15:34, 11.84it/s]\u001b[2m\u001b[36m(raylet)\u001b[0m Spilled 262836 MiB, 1233 objects, write throughput 2108 MiB/s.\n",
      "initializing:  16%|██████████▎                                                     | 1935/12080 [04:12<13:12, 12.80it/s]\u001b[2m\u001b[36m(raylet)\u001b[0m Spilled 526700 MiB, 2458 objects, write throughput 2123 MiB/s.\n",
      "initializing:  32%|████████████████████▍                                           | 3867/12080 [08:32<09:11, 14.90it/s]\u001b[2m\u001b[36m(raylet)\u001b[0m Spilled 1049664 MiB, 4871 objects, write throughput 2082 MiB/s.\n",
      "initializing:  64%|████████████████████████████████████████▋                       | 7683/12080 [17:19<09:17,  7.88it/s]\u001b[2m\u001b[36m(raylet)\u001b[0m Spilled 2097723 MiB, 9749 objects, write throughput 2048 MiB/s.\n",
      "initializing: 100%|███████████████████████████████████████████████████████████████| 12080/12080 [27:34<00:00,  7.30it/s]\n",
      "Running using 12 cores: 100%|████████████████████████████████████████████████████| 12080/12080 [01:16<00:00, 158.09it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-06-05 10:53:49,860 TF2G         INFO     Took 1733.723562002182 seconds\n",
      "2024-06-05 10:53:49,862 TF2G         INFO     Adding correlation coefficients to adjacencies.\n",
      "2024-06-05 10:54:23,292 TF2G         INFO     Warning: adding TFs as their own target to adjecencies matrix. Importance values will be max + 1e-05\n",
      "2024-06-05 10:54:26,850 TF2G         INFO     Adding importance x rho scores to adjacencies.\n",
      "2024-06-05 10:54:26,859 TF2G         INFO     Took 36.99719214439392 seconds\n"
     ]
    }
   ],
   "source": [
    "from scenicplus.TF_to_gene import *\n",
    "calculate_TFs_to_genes_relationships(scplus_obj,\n",
    "                    tf_file = tf_file, # list of TF names\n",
    "                    ray_n_cpu = 12,\n",
    "                    method = 'GBM',\n",
    "                    _temp_dir = tmp_dir, \n",
    "                    key= 'TF2G_adj')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "88b5c254-3812-42b9-a7e0-5484325befaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save\n",
    "import pickle\n",
    "\n",
    "with open(os.path.join(work_dir, 'scenicplus/scplus_obj_hocomocov11.pkl'), 'wb') as f:\n",
    "  pickle.dump(scplus_obj, f)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "538db25a-4798-4456-b1be-bf7b5625b9b2",
   "metadata": {},
   "source": [
    "Build eGRNs\n",
    "    uses a GSEA recovery approach\n",
    "    ranks are based on TF to gene importance\n",
    "    gene sets are derived from enhancer to gene relationships and unfiltered cistromes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8e485cc6-86b4-4c93-a0a4-174bd52acdbf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-06-05 11:00:28,922 GSEA         INFO     Thresholding region to gene relationships\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-06-05 11:00:32,735\tINFO worker.py:1612 -- Started a local Ray instance. View the dashboard at \u001b[1m\u001b[32m127.0.0.1:8265 \u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-06-05 11:03:31,798 GSEA         INFO     Subsetting TF2G adjacencies for TF with motif.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-06-05 11:03:35,938\tINFO worker.py:1612 -- Started a local Ray instance. View the dashboard at \u001b[1m\u001b[32m127.0.0.1:8265 \u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-06-05 11:03:36,852 GSEA         INFO     Running GSEA...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(_ray_run_gsea_for_e_module pid=3519)\u001b[0m /home/devoes/miniconda3/envs/SCENICplus_SCA/lib/python3.10/site-packages/gseapy/algorithm.py:71: RuntimeWarning: divide by zero encountered in divide\n",
      "\u001b[2m\u001b[36m(_ray_run_gsea_for_e_module pid=3519)\u001b[0m   norm_tag =  1.0/sum_correl_tag\n",
      "\u001b[2m\u001b[36m(_ray_run_gsea_for_e_module pid=3519)\u001b[0m /home/devoes/miniconda3/envs/SCENICplus_SCA/lib/python3.10/site-packages/gseapy/algorithm.py:74: RuntimeWarning: invalid value encountered in multiply\n",
      "\u001b[2m\u001b[36m(_ray_run_gsea_for_e_module pid=3519)\u001b[0m   RES = np.cumsum(tag_indicator * correl_vector * norm_tag - no_tag_indicator * norm_no_tag, axis=axis)\n",
      "\u001b[2m\u001b[36m(_ray_run_gsea_for_e_module pid=3523)\u001b[0m /home/devoes/miniconda3/envs/SCENICplus_SCA/lib/python3.10/site-packages/gseapy/algorithm.py:71: RuntimeWarning: divide by zero encountered in divide\u001b[32m [repeated 6x across cluster] (Ray deduplicates logs by default. Set RAY_DEDUP_LOGS=0 to disable log deduplication, or see https://docs.ray.io/en/master/ray-observability/ray-logging.html#log-deduplication for more options.)\u001b[0m\n",
      "\u001b[2m\u001b[36m(_ray_run_gsea_for_e_module pid=3523)\u001b[0m   norm_tag =  1.0/sum_correl_tag\u001b[32m [repeated 6x across cluster]\u001b[0m\n",
      "\u001b[2m\u001b[36m(_ray_run_gsea_for_e_module pid=3523)\u001b[0m /home/devoes/miniconda3/envs/SCENICplus_SCA/lib/python3.10/site-packages/gseapy/algorithm.py:74: RuntimeWarning: invalid value encountered in multiply\u001b[32m [repeated 6x across cluster]\u001b[0m\n",
      "\u001b[2m\u001b[36m(_ray_run_gsea_for_e_module pid=3523)\u001b[0m   RES = np.cumsum(tag_indicator * correl_vector * norm_tag - no_tag_indicator * norm_no_tag, axis=axis)\u001b[32m [repeated 6x across cluster]\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-06-05 11:05:15,318 GSEA         INFO     Subsetting on adjusted pvalue: 1, minimal NES: 0 and minimal leading edge genes 10\n",
      "2024-06-05 11:05:15,644 GSEA         INFO     Merging eRegulons\n",
      "2024-06-05 11:05:15,700 GSEA         INFO     Storing eRegulons in .uns[eRegulons_importance].\n"
     ]
    }
   ],
   "source": [
    "from scenicplus.grn_builder.gsea_approach import build_grn\n",
    "\n",
    "build_grn(scplus_obj,\n",
    "         min_target_genes = 10,\n",
    "         adj_pval_thr = 1,\n",
    "         min_regions_per_gene = 0,\n",
    "         quantiles = (0.85, 0.90), # 0.95?\n",
    "         top_n_regionTogenes_per_gene = (5, 10, 15),\n",
    "         top_n_regionTogenes_per_region = (),\n",
    "         binarize_using_basc = True,\n",
    "         rho_dichotomize_tf2g = True,\n",
    "         rho_dichotomize_r2g = True,\n",
    "         rho_dichotomize_eregulon = True,\n",
    "         rho_threshold = 0.05,\n",
    "         keep_extended_motif_annot = True,\n",
    "         merge_eRegulons = True,\n",
    "         order_regions_to_genes_by = 'importance',\n",
    "         order_TFs_to_genes_by = 'importance',\n",
    "         key_added = 'eRegulons_importance',\n",
    "         cistromes_key = 'Unfiltered',\n",
    "         disable_tqdm = True, #If running in notebook, set to True\n",
    "         ray_n_cpu = 12,\n",
    "         _temp_dir = tmp_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c79046ca-5a4d-4cb2-a041-e04395d7b4c3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import dill\n",
    "with open(os.path.join(work_dir, 'scenicplus/scplus_obj_hocomocov11.pkl'), 'wb') as f:\n",
    "  dill.dump(scplus_obj, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "248f2a89-f41f-4c90-98a9-8f51bfe09c9c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SCENICplus_SCA",
   "language": "python",
   "name": "scenicplus_sca"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
