{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2ae553f4-f427-4526-9c2b-4052da829cac",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# load candidate enhancers determined from B16CD8TIL-multiome_SCENICplus_scATACpp_cistopic\n",
    "import os\n",
    "import pickle\n",
    "work_dir = '/Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART'\n",
    "tmp_dir = '/scratch2/devoes/tmp'\n",
    "\n",
    "region_bin_topics_otsu = pickle.load(open(os.path.join(work_dir, 'ATAC/candidate_enhancers/region_bin_topics_otsu.pkl'), 'rb'))\n",
    "region_bin_topics_top3k = pickle.load(open(os.path.join(work_dir, 'ATAC/candidate_enhancers/region_bin_topics_top3k.pkl'), 'rb'))\n",
    "markers_dict = pickle.load(open(os.path.join(work_dir, 'ATAC/candidate_enhancers/markers_dict.pkl'), 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "49d9df95-5ba8-4e7e-8679-f48dc74e8aa3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# format as dictionary of PyRanges objects\n",
    "import pyranges as pr\n",
    "from pycistarget.utils import region_names_to_coordinates\n",
    "region_sets = {}\n",
    "region_sets['topics_otsu'] = {}\n",
    "region_sets['topics_top_3'] = {}\n",
    "region_sets['DARs'] = {}\n",
    "for topic in region_bin_topics_otsu.keys():\n",
    "    regions = region_bin_topics_otsu[topic].index[region_bin_topics_otsu[topic].index.str.startswith('chr')] #only keep regions on known chromosomes\n",
    "    region_sets['topics_otsu'][topic] = pr.PyRanges(region_names_to_coordinates(regions))\n",
    "for topic in region_bin_topics_top3k.keys():\n",
    "    regions = region_bin_topics_top3k[topic].index[region_bin_topics_top3k[topic].index.str.startswith('chr')] #only keep regions on known chromosomes\n",
    "    region_sets['topics_top_3'][topic] = pr.PyRanges(region_names_to_coordinates(regions))\n",
    "for DAR in markers_dict.keys():\n",
    "    #print(DAR)\n",
    "    regions = markers_dict[DAR].index[markers_dict[DAR].index.str.startswith('chr')] #only keep regions on known chromosomes\n",
    "    if len(regions) > 0: # no DARs for 10, 14, 2, 3, 8 ,9\n",
    "        region_sets['DARs'][DAR] = pr.PyRanges(region_names_to_coordinates(regions))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "99f336da-f25b-49f3-8f42-e0b8d10a327d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "topics_otsu: dict_keys(['Topic1', 'Topic2', 'Topic3', 'Topic4', 'Topic5', 'Topic6', 'Topic7', 'Topic8', 'Topic9', 'Topic10', 'Topic11', 'Topic12', 'Topic13', 'Topic14', 'Topic15', 'Topic16', 'Topic17', 'Topic18', 'Topic19', 'Topic20', 'Topic21', 'Topic22', 'Topic23', 'Topic24', 'Topic25', 'Topic26', 'Topic27', 'Topic28', 'Topic29', 'Topic30', 'Topic31', 'Topic32', 'Topic33', 'Topic34', 'Topic35', 'Topic36', 'Topic37', 'Topic38', 'Topic39', 'Topic40', 'Topic41', 'Topic42', 'Topic43', 'Topic44', 'Topic45', 'Topic46', 'Topic47', 'Topic48'])\n",
      "topics_top_3: dict_keys(['Topic1', 'Topic2', 'Topic3', 'Topic4', 'Topic5', 'Topic6', 'Topic7', 'Topic8', 'Topic9', 'Topic10', 'Topic11', 'Topic12', 'Topic13', 'Topic14', 'Topic15', 'Topic16', 'Topic17', 'Topic18', 'Topic19', 'Topic20', 'Topic21', 'Topic22', 'Topic23', 'Topic24', 'Topic25', 'Topic26', 'Topic27', 'Topic28', 'Topic29', 'Topic30', 'Topic31', 'Topic32', 'Topic33', 'Topic34', 'Topic35', 'Topic36', 'Topic37', 'Topic38', 'Topic39', 'Topic40', 'Topic41', 'Topic42', 'Topic43', 'Topic44', 'Topic45', 'Topic46', 'Topic47', 'Topic48'])\n",
      "DARs: dict_keys(['0', '1', '11', '12', '13', '3', '4', '5', '6', '7'])\n"
     ]
    }
   ],
   "source": [
    "for key in region_sets.keys():\n",
    "    print(f'{key}: {region_sets[key].keys()}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ec4fdf9d-c62d-4bfc-bca9-84bcecc4ff91",
   "metadata": {},
   "outputs": [],
   "source": [
    "#set paths for pycistarget databases\n",
    "rankings_db = os.path.join(work_dir, 'cistarget_database/SD031_B16_CART.regions_vs_motifs.rankings.feather')\n",
    "scores_db =  os.path.join(work_dir, 'cistarget_database/SD031_B16_CART.regions_vs_motifs.scores.feather')\n",
    "motif_annotation = '/Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART/cistarget_database/HOCOMOCOv11_motifs.tbl'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f09c96a6-4d51-4b92-91c6-6bd453615a26",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-06-04 17:57:34,217 pycisTarget_wrapper INFO     /Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART/motifs_no-promoter folder already exists.\n",
      "2024-06-04 17:57:34,744 pycisTarget_wrapper INFO     Loading cisTarget database for topics_otsu\n",
      "2024-06-04 17:57:34,746 cisTarget    INFO     Reading cisTarget database\n",
      "2024-06-04 17:59:06,635 pycisTarget_wrapper INFO     Running cisTarget for topics_otsu\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=16118)\u001b[0m 2024-06-04 17:59:13,116 cisTarget    INFO     Running cisTarget for Topic1 which has 9306 regions\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=16118)\u001b[0m 2024-06-04 17:59:17,571 cisTarget    INFO     No enriched motifs found for Topic1\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=16118)\u001b[0m 2024-06-04 17:59:17,857 cisTarget    INFO     No enriched motifs found for Topic9\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=16121)\u001b[0m 2024-06-04 17:59:18,361 cisTarget    INFO     Annotating motifs for Topic2\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=16118)\u001b[0m 2024-06-04 17:59:18,004 cisTarget    INFO     Running cisTarget for Topic10 which has 4088 regions\u001b[32m [repeated 9x across cluster]\u001b[0m\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=16115)\u001b[0m 2024-06-04 17:59:18,303 cisTarget    INFO     No enriched motifs found for Topic3\n",
      "2024-06-04 17:59:32,818 cisTarget    INFO     Done!\n",
      "2024-06-04 17:59:32,819 pycisTarget_wrapper INFO     /Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART/motifs_no-promoter/CTX_topics_otsu_All folder already exists.\n",
      "2024-06-04 17:59:33,001 pycisTarget_wrapper INFO     Running cisTarget without promoters for topics_otsu\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=16905)\u001b[0m 2024-06-04 18:00:37,898 cisTarget    INFO     Running cisTarget for Topic1 which has 8285 regions\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=16905)\u001b[0m 2024-06-04 18:00:42,577 cisTarget    INFO     No enriched motifs found for Topic1\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=16905)\u001b[0m 2024-06-04 18:00:42,856 cisTarget    INFO     No enriched motifs found for Topic9\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=16905)\u001b[0m 2024-06-04 18:00:42,993 cisTarget    INFO     Running cisTarget for Topic10 which has 3864 regions\u001b[32m [repeated 9x across cluster]\u001b[0m\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=16903)\u001b[0m 2024-06-04 18:00:43,361 cisTarget    INFO     Annotating motifs for Topic2\n",
      "2024-06-04 18:00:53,553 cisTarget    INFO     Done!\n",
      "2024-06-04 18:00:53,555 pycisTarget_wrapper INFO     /Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART/motifs_no-promoter/CTX_topics_otsu_No_promoters folder already exists.\n",
      "2024-06-04 18:00:53,731 pycisTarget_wrapper INFO     Running DEM for topics_otsu\n",
      "2024-06-04 18:00:53,732 DEM          INFO     Reading DEM database\n",
      "2024-06-04 18:02:26,120 DEM          INFO     Creating contrast groups\n",
      "\u001b[2m\u001b[36m(DEM_internal_ray pid=17995)\u001b[0m 2024-06-04 18:04:12,967 DEM          INFO     Computing DEM for Topic2\n",
      "\u001b[2m\u001b[36m(DEM_internal_ray pid=17998)\u001b[0m 2024-06-04 18:04:18,624 DEM          INFO     Computing DEM for Topic20\u001b[32m [repeated 19x across cluster]\u001b[0m\n",
      "\u001b[2m\u001b[36m(DEM_internal_ray pid=17996)\u001b[0m 2024-06-04 18:04:23,850 DEM          INFO     Computing DEM for Topic35\u001b[32m [repeated 15x across cluster]\u001b[0m\n",
      "\u001b[2m\u001b[36m(DEM_internal_ray pid=17997)\u001b[0m 2024-06-04 18:04:28,809 DEM          INFO     Computing DEM for Topic45\u001b[32m [repeated 10x across cluster]\u001b[0m\n",
      "\u001b[2m\u001b[36m(DEM_internal_ray pid=17993)\u001b[0m 2024-06-04 18:04:29,593 DEM          INFO     Computing DEM for Topic48\u001b[32m [repeated 3x across cluster]\u001b[0m\n",
      "2024-06-04 18:04:41,641 DEM          INFO     Forming cistromes\n",
      "2024-06-04 18:04:43,430 DEM          INFO     Done!\n",
      "2024-06-04 18:04:43,432 pycisTarget_wrapper INFO     /Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART/motifs_no-promoter/DEM_topics_otsu_All folder already exists.\n",
      "2024-06-04 18:04:43,720 pycisTarget_wrapper INFO     Running DEM without promoters for topics_otsu\n",
      "2024-06-04 18:05:41,914 DEM          INFO     Creating contrast groups\n",
      "\u001b[2m\u001b[36m(DEM_internal_ray pid=18794)\u001b[0m 2024-06-04 18:06:01,354 DEM          INFO     Computing DEM for Topic2\n",
      "\u001b[2m\u001b[36m(DEM_internal_ray pid=18795)\u001b[0m 2024-06-04 18:06:07,254 DEM          INFO     Computing DEM for Topic20\u001b[32m [repeated 18x across cluster]\u001b[0m\n",
      "\u001b[2m\u001b[36m(DEM_internal_ray pid=18794)\u001b[0m 2024-06-04 18:06:07,937 DEM          INFO     Computing DEM for Topic21\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "2024-06-04 18:06:25,914 DEM          INFO     Forming cistromes\n",
      "2024-06-04 18:06:27,740 DEM          INFO     Done!\n",
      "2024-06-04 18:06:27,742 pycisTarget_wrapper INFO     /Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART/motifs_no-promoter/DEM_topics_otsu_No_promoters folder already exists.\n",
      "2024-06-04 18:06:28,032 pycisTarget_wrapper INFO     Loading cisTarget database for topics_top_3\n",
      "2024-06-04 18:06:28,033 cisTarget    INFO     Reading cisTarget database\n",
      "2024-06-04 18:07:50,571 pycisTarget_wrapper INFO     Running cisTarget for topics_top_3\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=19694)\u001b[0m 2024-06-04 18:07:56,329 cisTarget    INFO     Running cisTarget for Topic1 which has 2997 regions\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=19701)\u001b[0m 2024-06-04 18:08:01,122 cisTarget    INFO     No enriched motifs found for Topic3\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=19701)\u001b[0m 2024-06-04 18:08:01,276 cisTarget    INFO     No enriched motifs found for Topic9\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=19694)\u001b[0m 2024-06-04 18:08:01,474 cisTarget    INFO     Annotating motifs for Topic1\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=19701)\u001b[0m 2024-06-04 18:08:01,357 cisTarget    INFO     Running cisTarget for Topic10 which has 2997 regions\u001b[32m [repeated 9x across cluster]\u001b[0m\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=19694)\u001b[0m 2024-06-04 18:08:01,660 cisTarget    INFO     Getting cistromes for Topic1\n",
      "2024-06-04 18:08:15,610 cisTarget    INFO     Done!\n",
      "2024-06-04 18:08:15,612 pycisTarget_wrapper INFO     /Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART/motifs_no-promoter/CTX_topics_top_3_All folder already exists.\n",
      "2024-06-04 18:08:15,802 pycisTarget_wrapper INFO     Running cisTarget without promoters for topics_top_3\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=20598)\u001b[0m 2024-06-04 18:09:02,254 cisTarget    INFO     Running cisTarget for Topic1 which has 2636 regions\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=20599)\u001b[0m 2024-06-04 18:09:03,095 cisTarget    INFO     Running cisTarget for Topic8 which has 2737 regions\u001b[32m [repeated 7x across cluster]\u001b[0m\n",
      "2024-06-04 18:09:41,923 cisTarget    INFO     Done!\n",
      "2024-06-04 18:09:41,925 pycisTarget_wrapper INFO     /Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART/motifs_no-promoter/CTX_topics_top_3_No_promoters folder already exists.\n",
      "2024-06-04 18:09:42,130 pycisTarget_wrapper INFO     Running DEM for topics_top_3\n",
      "2024-06-04 18:09:42,131 DEM          INFO     Reading DEM database\n",
      "2024-06-04 18:11:12,908 DEM          INFO     Creating contrast groups\n",
      "\u001b[2m\u001b[36m(DEM_internal_ray pid=21629)\u001b[0m 2024-06-04 18:12:20,019 DEM          INFO     Computing DEM for Topic1\n",
      "\u001b[2m\u001b[36m(DEM_internal_ray pid=21625)\u001b[0m 2024-06-04 18:12:23,406 DEM          INFO     Computing DEM for Topic15\u001b[32m [repeated 14x across cluster]\u001b[0m\n",
      "2024-06-04 18:12:48,257 DEM          INFO     Forming cistromes\n",
      "2024-06-04 18:12:49,901 DEM          INFO     Done!\n",
      "2024-06-04 18:12:49,902 pycisTarget_wrapper INFO     /Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART/motifs_no-promoter/DEM_topics_top_3_All folder already exists.\n",
      "2024-06-04 18:12:50,204 pycisTarget_wrapper INFO     Running DEM without promoters for topics_top_3\n",
      "2024-06-04 18:13:33,561 DEM          INFO     Creating contrast groups\n",
      "\u001b[2m\u001b[36m(DEM_internal_ray pid=22662)\u001b[0m 2024-06-04 18:13:49,499 DEM          INFO     Computing DEM for Topic1\n",
      "\u001b[2m\u001b[36m(DEM_internal_ray pid=22661)\u001b[0m 2024-06-04 18:13:52,917 DEM          INFO     Computing DEM for Topic18\u001b[32m [repeated 16x across cluster]\u001b[0m\n",
      "2024-06-04 18:14:25,336 DEM          INFO     Forming cistromes\n",
      "2024-06-04 18:14:26,581 DEM          INFO     Done!\n",
      "2024-06-04 18:14:26,583 pycisTarget_wrapper INFO     /Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART/motifs_no-promoter/DEM_topics_top_3_No_promoters folder already exists.\n",
      "2024-06-04 18:14:26,818 pycisTarget_wrapper INFO     Loading cisTarget database for DARs\n",
      "2024-06-04 18:14:26,819 cisTarget    INFO     Reading cisTarget database\n",
      "2024-06-04 18:14:47,463 pycisTarget_wrapper INFO     Running cisTarget for DARs\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=23693)\u001b[0m 2024-06-04 18:14:53,219 cisTarget    INFO     Running cisTarget for 1 which has 8220 regions\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=23690)\u001b[0m 2024-06-04 18:14:53,347 cisTarget    INFO     Running cisTarget for 5 which has 10272 regions\u001b[32m [repeated 7x across cluster]\u001b[0m\n",
      "2024-06-04 18:15:01,939 cisTarget    INFO     Done!\n",
      "2024-06-04 18:15:01,941 pycisTarget_wrapper INFO     /Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART/motifs_no-promoter/CTX_DARs_All folder already exists.\n",
      "2024-06-04 18:15:02,004 pycisTarget_wrapper INFO     Running cisTarget without promoters for DARs\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=24242)\u001b[0m 2024-06-04 18:15:13,495 cisTarget    INFO     Running cisTarget for 0 which has 1819 regions\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=24245)\u001b[0m 2024-06-04 18:15:18,408 cisTarget    INFO     Annotating motifs for 11\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=24245)\u001b[0m 2024-06-04 18:15:18,465 cisTarget    INFO     Getting cistromes for 11\n",
      "\u001b[2m\u001b[36m(ctx_internal_ray pid=24240)\u001b[0m 2024-06-04 18:15:13,652 cisTarget    INFO     Running cisTarget for 5 which has 9349 regions\u001b[32m [repeated 7x across cluster]\u001b[0m\n",
      "2024-06-04 18:15:21,965 cisTarget    INFO     Done!\n",
      "2024-06-04 18:15:21,967 pycisTarget_wrapper INFO     /Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART/motifs_no-promoter/CTX_DARs_No_promoters folder already exists.\n",
      "2024-06-04 18:15:22,031 pycisTarget_wrapper INFO     Running DEM for DARs\n",
      "2024-06-04 18:15:22,032 DEM          INFO     Reading DEM database\n",
      "2024-06-04 18:15:43,038 DEM          INFO     Creating contrast groups\n",
      "\u001b[2m\u001b[36m(DEM_internal_ray pid=24779)\u001b[0m 2024-06-04 18:15:56,390 DEM          INFO     Computing DEM for 0\n",
      "\u001b[2m\u001b[36m(DEM_internal_ray pid=24781)\u001b[0m 2024-06-04 18:15:57,450 DEM          INFO     Computing DEM for 7\u001b[32m [repeated 9x across cluster]\u001b[0m\n",
      "2024-06-04 18:16:04,199 DEM          INFO     Forming cistromes\n",
      "2024-06-04 18:16:04,463 DEM          INFO     Done!\n",
      "2024-06-04 18:16:04,465 pycisTarget_wrapper INFO     /Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART/motifs_no-promoter/DEM_DARs_All folder already exists.\n",
      "2024-06-04 18:16:04,520 pycisTarget_wrapper INFO     Running DEM without promoters for DARs\n",
      "2024-06-04 18:16:10,216 DEM          INFO     Creating contrast groups\n",
      "\u001b[2m\u001b[36m(DEM_internal_ray pid=25315)\u001b[0m 2024-06-04 18:16:16,611 DEM          INFO     Computing DEM for 11\n",
      "\u001b[2m\u001b[36m(DEM_internal_ray pid=25312)\u001b[0m 2024-06-04 18:16:17,739 DEM          INFO     Computing DEM for 6\u001b[32m [repeated 9x across cluster]\u001b[0m\n",
      "2024-06-04 18:16:23,331 DEM          INFO     Forming cistromes\n",
      "2024-06-04 18:16:23,599 DEM          INFO     Done!\n",
      "2024-06-04 18:16:23,602 pycisTarget_wrapper INFO     /Scottbrowne/members/smd/Projects/SD031/scenicplus/B16_CART/motifs_no-promoter/DEM_DARs_No_promoters folder already exists.\n",
      "2024-06-04 18:16:23,640 pycisTarget_wrapper INFO     Saving object\n",
      "2024-06-04 18:20:06,628 pycisTarget_wrapper INFO     Finished! Took 22.54017528692881 minutes\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists(os.path.join(work_dir, 'motifs_no-promoter')):\n",
    "    os.makedirs(os.path.join(work_dir, 'motifs_no-promoter'))\n",
    "    \n",
    "# run pycistarget wrapper\n",
    "# score and rank each region based on a motif collection and the region's score for each motif\n",
    "# \"The input regions are intersected with regions in the database (with at least 40% overlap). \n",
    "# cisTarget uses a recovery curve approach (for each motif), \n",
    "# in which a step is taken in the y-axis when as region in the motif ranking (x-axis) is found in the region set.\n",
    "# The Area Under the Curve for each motif is normalized based on the average AUC for all motifs and their standard deviation, \n",
    "# resulting in a Normalized Enrichment Score (NES) that is used to quantify the enrichment of a motif in a set of regions. \n",
    "# By default, motif that obtain a NES above 3.0 are kept.\"  https://pycistarget.readthedocs.io/en/latest/tools.html#cistarget \n",
    "from scenicplus.wrappers.run_pycistarget import run_pycistarget\n",
    "run_pycistarget(\n",
    "    region_sets = region_sets,\n",
    "    species = 'mus_musculus',\n",
    "    biomart_host = 'http://nov2020.archive.ensembl.org/',\n",
    "    save_path = os.path.join(work_dir, 'motifs_no-promoter'),\n",
    "    ctx_db_path = rankings_db,\n",
    "    dem_db_path = scores_db,\n",
    "    path_to_motif_annotations = motif_annotation,\n",
    "    run_without_promoters = True, \n",
    "    n_cpu = 8,\n",
    "    _temp_dir = os.path.join(tmp_dir, 'ray_spill'),\n",
    "    annotation_version = 'HOCOMOCOv11',\n",
    "    )"
   ]
  },
  {
   "cell_type": "raw",
   "id": "df3a3800-e730-4dc2-971a-eca448ac3438",
   "metadata": {},
   "source": [
    "Explore results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e81f26cf-6908-4922-b06d-6c394398607a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import dill\n",
    "menr = dill.load(open(os.path.join(work_dir, 'motifs_no-promoter/menr.pkl'), 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cbc868ef-3d3e-4eac-bd31-65d0d9b4bf83",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th style=\"min-width: 80px;\"></th>\n",
       "      <th style=\"min-width: 80px;\">Logo</th>\n",
       "      <th style=\"min-width: 80px;\">Contrast</th>\n",
       "      <th style=\"min-width: 80px;\">Direct_annot</th>\n",
       "      <th style=\"min-width: 80px;\">Orthology_annot</th>\n",
       "      <th style=\"min-width: 80px;\">Log2FC</th>\n",
       "      <th style=\"min-width: 80px;\">Adjusted_pval</th>\n",
       "      <th style=\"min-width: 80px;\">Mean_fg</th>\n",
       "      <th style=\"min-width: 80px;\">Mean_bg</th>\n",
       "      <th style=\"min-width: 80px;\">Motif_hit_thr</th>\n",
       "      <th style=\"min-width: 80px;\">Motif_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FOS_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/FOS_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Fos</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.526217</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.554405</td>\n",
       "      <td>0.617021</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1502.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOSL1_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/FOSL1_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Fosl1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.425178</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.076396</td>\n",
       "      <td>0.758971</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1782.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JUND_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/JUND_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Jund</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.973197</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.202226</td>\n",
       "      <td>1.070257</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1820.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOSL2_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/FOSL2_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Fosl2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.967752</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.273167</td>\n",
       "      <td>1.09244</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1835.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JUN_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/JUN_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Jun</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.811483</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.529927</td>\n",
       "      <td>1.290567</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1943.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JUNB_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/JUNB_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Junb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.74972</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.651789</td>\n",
       "      <td>1.383253</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2082.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NFE2_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/NFE2_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Nfe2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.747571</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.928845</td>\n",
       "      <td>0.574416</td>\n",
       "      <td>3.0</td>\n",
       "      <td>729.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BATF_MOUSE.H11MO.1.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/BATF_MOUSE.H11MO.1.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Batf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.638091</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.40724</td>\n",
       "      <td>1.094682</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1564.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOSB_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/FOSB_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Fosb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.55158</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.800166</td>\n",
       "      <td>1.63751</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2156.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATF3_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/ATF3_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Atf3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.489029</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.842672</td>\n",
       "      <td>1.725213</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2082.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BACH2_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/BACH2_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Bach2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.484571</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.522058</td>\n",
       "      <td>1.258625</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1493.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BACH1_MOUSE.H11MO.0.C</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/BACH1_MOUSE.H11MO.0.C.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Bach1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.304697</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.177943</td>\n",
       "      <td>0.881645</td>\n",
       "      <td>3.0</td>\n",
       "      <td>805.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BATF_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/BATF_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Batf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.279203</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.929634</td>\n",
       "      <td>1.207077</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1290.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAFK_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/MAFK_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Mafk</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.05028</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.979645</td>\n",
       "      <td>0.473046</td>\n",
       "      <td>3.0</td>\n",
       "      <td>354.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAFF_MOUSE.H11MO.1.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/MAFF_MOUSE.H11MO.1.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Maff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.048773</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.542579</td>\n",
       "      <td>1.229029</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1033.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NF2L2_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/NF2L2_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Nfe2l2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.010674</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.898561</td>\n",
       "      <td>1.438598</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1167.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAFG_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/MAFG_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Mafg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.985093</td>\n",
       "      <td>0.00047</td>\n",
       "      <td>0.510256</td>\n",
       "      <td>0.257778</td>\n",
       "      <td>3.0</td>\n",
       "      <td>188.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NF2L1_MOUSE.H11MO.0.C</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/NF2L1_MOUSE.H11MO.0.C.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Nfe2l1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.930578</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.787</td>\n",
       "      <td>1.462194</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1109.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAFK_MOUSE.H11MO.1.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/MAFK_MOUSE.H11MO.1.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Mafk</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.807623</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.278216</td>\n",
       "      <td>1.301596</td>\n",
       "      <td>3.0</td>\n",
       "      <td>901.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RORG_MOUSE.H11MO.0.B</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/RORG_MOUSE.H11MO.0.B.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Rorc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.79801</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.303551</td>\n",
       "      <td>1.324869</td>\n",
       "      <td>3.0</td>\n",
       "      <td>953.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PAX2_MOUSE.H11MO.0.D</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/PAX2_MOUSE.H11MO.0.D.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Pax2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.765205</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.257329</td>\n",
       "      <td>0.739775</td>\n",
       "      <td>3.0</td>\n",
       "      <td>334.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RUNX1_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/RUNX1_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Runx1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.650272</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.379649</td>\n",
       "      <td>1.516218</td>\n",
       "      <td>3.0</td>\n",
       "      <td>987.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NR1D2_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/NR1D2_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Nr1d2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.636214</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.907568</td>\n",
       "      <td>1.870727</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1202.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RUNX3_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/RUNX3_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Runx3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.589309</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.47819</td>\n",
       "      <td>1.647157</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1069.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CREB1_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/CREB1_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Creb1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.544785</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.179133</td>\n",
       "      <td>0.808288</td>\n",
       "      <td>3.0</td>\n",
       "      <td>253.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RUNX2_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/RUNX2_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Runx2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.54119</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.628257</td>\n",
       "      <td>1.806148</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1119.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CEBPG_MOUSE.H11MO.0.B</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/CEBPG_MOUSE.H11MO.0.B.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Cebpg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.526326</td>\n",
       "      <td>0.002237</td>\n",
       "      <td>0.528063</td>\n",
       "      <td>0.366645</td>\n",
       "      <td>3.0</td>\n",
       "      <td>149.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NFKB1_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/NFKB1_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Nfkb1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.516201</td>\n",
       "      <td>0.001542</td>\n",
       "      <td>1.158711</td>\n",
       "      <td>0.810183</td>\n",
       "      <td>3.0</td>\n",
       "      <td>424.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TF65_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/TF65_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic2</td>\n",
       "      <td>Rela</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.511534</td>\n",
       "      <td>0.000742</td>\n",
       "      <td>1.418554</td>\n",
       "      <td>0.995082</td>\n",
       "      <td>3.0</td>\n",
       "      <td>537.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "menr['DEM_topics_otsu_All'].DEM_results('Topic2') #2,4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "25147979-bf66-4d39-a259-9386a495e06d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th style=\"min-width: 80px;\"></th>\n",
       "      <th style=\"min-width: 80px;\">Logo</th>\n",
       "      <th style=\"min-width: 80px;\">Contrast</th>\n",
       "      <th style=\"min-width: 80px;\">Direct_annot</th>\n",
       "      <th style=\"min-width: 80px;\">Orthology_annot</th>\n",
       "      <th style=\"min-width: 80px;\">Log2FC</th>\n",
       "      <th style=\"min-width: 80px;\">Adjusted_pval</th>\n",
       "      <th style=\"min-width: 80px;\">Mean_fg</th>\n",
       "      <th style=\"min-width: 80px;\">Mean_bg</th>\n",
       "      <th style=\"min-width: 80px;\">Motif_hit_thr</th>\n",
       "      <th style=\"min-width: 80px;\">Motif_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TF7L2_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/TF7L2_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Tcf7l2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.067834</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.481409</td>\n",
       "      <td>1.183718</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1008.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOS_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/FOS_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Fos</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.063158</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.333673</td>\n",
       "      <td>0.638274</td>\n",
       "      <td>3.0</td>\n",
       "      <td>520.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOSL1_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/FOSL1_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Fosl1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.062401</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.54959</td>\n",
       "      <td>0.741997</td>\n",
       "      <td>3.0</td>\n",
       "      <td>636.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TF7L1_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/TF7L1_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Tcf7l1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.962981</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.637887</td>\n",
       "      <td>1.353225</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1083.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JUND_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/JUND_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Jund</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.779909</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.839299</td>\n",
       "      <td>1.071214</td>\n",
       "      <td>3.0</td>\n",
       "      <td>676.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOSL2_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/FOSL2_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Fosl2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.765584</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.773321</td>\n",
       "      <td>1.043094</td>\n",
       "      <td>3.0</td>\n",
       "      <td>671.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BATF_MOUSE.H11MO.1.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/BATF_MOUSE.H11MO.1.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Batf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.721925</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.64484</td>\n",
       "      <td>0.997246</td>\n",
       "      <td>3.0</td>\n",
       "      <td>616.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TFE3_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/TFE3_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Tfe3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.704043</td>\n",
       "      <td>0.000333</td>\n",
       "      <td>0.751139</td>\n",
       "      <td>0.461086</td>\n",
       "      <td>3.0</td>\n",
       "      <td>203.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATF4_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/ATF4_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Atf4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.702804</td>\n",
       "      <td>0.038198</td>\n",
       "      <td>0.28744</td>\n",
       "      <td>0.176597</td>\n",
       "      <td>3.0</td>\n",
       "      <td>79.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JUNB_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/JUNB_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Junb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.698533</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.095639</td>\n",
       "      <td>1.29133</td>\n",
       "      <td>3.0</td>\n",
       "      <td>802.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JUN_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/JUN_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Jun</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.689566</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.020794</td>\n",
       "      <td>1.252974</td>\n",
       "      <td>3.0</td>\n",
       "      <td>743.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CEBPG_MOUSE.H11MO.0.B</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/CEBPG_MOUSE.H11MO.0.B.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Cebpg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.621307</td>\n",
       "      <td>0.00553</td>\n",
       "      <td>0.515575</td>\n",
       "      <td>0.335166</td>\n",
       "      <td>3.0</td>\n",
       "      <td>141.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ETV2_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/ETV2_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Etv2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.603794</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.917517</td>\n",
       "      <td>1.261767</td>\n",
       "      <td>3.0</td>\n",
       "      <td>785.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RUNX1_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/RUNX1_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Runx1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.577374</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.153542</td>\n",
       "      <td>1.443266</td>\n",
       "      <td>3.0</td>\n",
       "      <td>846.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RUNX3_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/RUNX3_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Runx3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.570726</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.340729</td>\n",
       "      <td>1.575961</td>\n",
       "      <td>3.0</td>\n",
       "      <td>952.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOSB_MOUSE.H11MO.0.A</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/FOSB_MOUSE.H11MO.0.A.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Fosb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.570553</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.284498</td>\n",
       "      <td>1.538287</td>\n",
       "      <td>3.0</td>\n",
       "      <td>854.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEIS2_MOUSE.H11MO.0.B</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/MEIS2_MOUSE.H11MO.0.B.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Meis2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.559904</td>\n",
       "      <td>0.004638</td>\n",
       "      <td>0.572982</td>\n",
       "      <td>0.388681</td>\n",
       "      <td>3.0</td>\n",
       "      <td>207.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EVI1_MOUSE.H11MO.0.B</th>\n",
       "      <td><img src=\"https://motifcollections.aertslab.org/HOCOMOCOv11/logos/EVI1_MOUSE.H11MO.0.B.png\" width=\"200\" ></td>\n",
       "      <td>Topic4</td>\n",
       "      <td>Mecom</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.524393</td>\n",
       "      <td>0.036508</td>\n",
       "      <td>0.335511</td>\n",
       "      <td>0.233264</td>\n",
       "      <td>3.0</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "menr['DEM_topics_otsu_All'].DEM_results('Topic4')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3542ddec-74e7-4752-b4ff-fda8066332bc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SCENICplus_SCA",
   "language": "python",
   "name": "scenicplus_sca"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
